### DISCUSSION TITLE
Improving the NoiseModelling OSM Import WPS Script

### DISCUSSION THREAD

**Discussion by IsotoCedex** [Universite-Gustave-Eiffel/NoiseModelling#715] [Ideas]
Below is an explanation of the code changes that were made:

**1. **Preventing NumberFormatException when parsing numeric values:****

> Previous issue:
> The script used expressions like:
> 
> ```
> Double.parseDouble(tag.getValue().replaceAll("[^0-9]+", ""))
> ```
> 
> If the string contained no digits, this resulted in an empty string, causing a NumberFormatException.
> 
> Solution implemented:
> Before parsing, we now check if the string is not empty:
> 
> ```
> String rawVal = tag.getValue().replaceAll("[^0-9]+", "")
> if (!rawVal.isEmpty()) {
>     height = Double.parseDouble(rawVal);
> }
> 
> ```
> This ensures that if no digits are found, the parsing step is skipped and a default value remains unchanged.
> 
> This pattern has been applied wherever numeric parsing occurs, such as for building heights (building:levels, height) and road maxspeed.

**2. Introducing the new eliminateNoTrafficRoads parameter:**

> New functionality:
> A new boolean input parameter was added:
> 
> ```
> eliminateNoTrafficRoads : [
>     name       : 'Eliminate no traffic roads',
>     title      : 'Eliminate no traffic roads',
>     description: '... (explanation of the road types that remain if enabled) ...',
>     type       : Boolean.class
> ]
> ```
> 
> Logic applied:
> If eliminateNoTrafficRoads is activated, after processing the OSM data, the script filters the handler.roads list to keep only roads whose type is included in a predefined set of allowed types. For example:
> 
> ```
> if (eliminateNoTrafficRoads && !ignoreRoads) {
>     def validRoadTypes = ["bus_guideway", "busway", "living_street", "motorway",
>                           "motorway_link", "primary", "primary_link", "raceway",
>                           "residential", "road", "secondary", "secondary_link",
>                           "service", "tertiary", "tertiary_link", "trunk",
>                           "trunk_link", "unclassified", "rest_area", "traffic_calming",
>                           "traffic_island"]
>     handler.roads = handler.roads.findAll { validRoadTypes.contains(it.type) }
> }
> ```
> 
> With this addition, enabling the parameter ensures that only roads considered "traffic-relevant" are included in the final output. If it’s disabled, all roads are processed as before.

**3. Maintaining Backward Compatibility:**

> If the new option is not activated, the script behaves exactly as it did previously, ensuring that users who prefer the original functionality do not need to alter their workflow.

**4. Improved Documentation:**

> We have updated the input descriptions and inline comments to make it clearer what each parameter does. While this doesn’t change the logic, it provides better guidance for users, explaining how each parameter influences the output.

**In summary:**

> The primary code changes focus on adding robustness to numeric parsing to prevent runtime errors and introducing a new filtering option to selectively include only certain road types. As a result, the script is now more resilient to incomplete data and offers greater flexibility, all while preserving the original functionality for users who do not wish to change their existing processes.

**The modified code.**

[Import_OSM.zip](https://github.com/user-attachments/files/18051850/Import_OSM.zip)

**Example**

All roads (red) and filtered roads (blue) in La Gomera (Canary Islands):

![image](https://github.com/user-attachments/assets/88ce7a2e-3db4-4ca1-bfd3-ba238d33a5c2)

**Removed Road Types and Their Descriptions**

- abandoned: Roads that are no longer in use and have been abandoned.
- bridleway: Paths primarily intended for horseback riding.
- bus_stop: Bus stops.
- construction: Roads under construction and not open to traffic.
- corridor: Indoor hallways in buildings or urban areas.
- cycleway: Routes dedicated to bicycles.
- demolished: Roads that have been demolished.
- disused: Roads no longer in use, but still physically exist.
- elevator: Elevators.
- emergency_bay: Emergency bays on roads, for emergency stops.
- escape: Escape lanes for vehicles in case of emergency.
- footway: Paths intended for pedestrians.
- ladder: Ladders.
- path: Various types of paths, generally not allowing motorized traffic.
- pedestrian: Pedestrian areas.
- platform: Platforms, generally associated with public transportation.
- proposed: Proposed roads, planned but not yet built.
- propossed: Typographical error of "proposed."
- razed: Roads that have been completely removed.
- ser: A non-standard OSM classification, possibly a typographical error.
- steps: Stairs.
- track: Access roads, generally in rural or agricultural areas, often mountainous and resembling trails.
- turntable: Turntables, usually for trains.
- via_ferrata: Mountain trails equipped with cables and iron steps.
- yes: A non-standard OSM classification, possibly a typographical error.
- services: Roadside service areas.

**Comment by pierromond** [comment]
Thank you @IsotoCedex ! could you do a Pull Request (in this way you will keep the authorship of the changes).

**Comment by IsotoCedex** [comment]
I will try... :)

**Comment by pierromond** [comment]
If you want we can do a zoom call to show you !

**Comment by IsotoCedex** [comment]
Hello @pierromond,
I think i have done it.
https://github.com/Universite-Gustave-Eiffel/NoiseModelling/commit/a33f99128f421eae5fb2b655c0b16092ab38ce6a

**Comment by pierromond** [comment]
Perfect, you succeeded to do the commit ! Then you can ask for a pull request to the 'NoiseModelling 4.X' branch of the 'Universite-Gustave-Eiffel' remote. Github proposes you to do so on the main page normally. 
https://docs.github.com/fr/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request

**Comment by IsotoCedex** [comment]
Thank you @pierromond 
Here it is https://github.com/Universite-Gustave-Eiffel/NoiseModelling/pull/720
